server/lib/models/matchmaking-times.ts

```diff
@@ -0,0 +1,132 @@
+import db from '../db'
+import sql from 'sql-template-strings'
+import { MatchmakingType } from '../../../common/matchmaking'
+
+export interface MatchmakingTime {
+  id: string
+  type: MatchmakingType
+  startDate: number
+  enabled: boolean
+}
+
+function convertFromDb(props: {
+  /* eslint-disable camelcase */
+  id: string
+  matchmaking_type: MatchmakingType
+  start_date: number
+  enabled: boolean
+  /* eslint-enable camelcase */
+}): MatchmakingTime {
+  return {
+    id: props.id,
+    type: props.matchmaking_type,
+    startDate: +props.start_date,
```

Comment:
Hmm, I'm not sure; it's definitely not a number so type on `props` was incorrect. It looks like a `Date`, but I don't know in which formatting.

I think returning it as a `number` makes more sense, since users usually create a new date out of it to format it to their liking (I'm not sure if this `Date` that's returned by postgres could be directly formatted). Also, I feel like comparing different dates (which is done in multiple places) is easier in their number form instead of their formatted `Date` form, but I could be wrong about this?

Reply:
sorry, I should have been more clear in my language. By "users of this API", I mean "callers of this function", not "users of the HTTP API that eventually calls into this function". `pg` returns `Date` objects for date columns, and those objects have a time zone associated with them, so by converting them to a number, you're throwing that information away. I understand that you eventually want to show these as a String on the client, but generally you want to keep data in it's most universal form until you need it in some other, specific form. e.g. if we wanted to display a number rounded to 2 decimal places on the client, we wouldn't do that rounding until the display layer, our database layer would return the full, unrounded form.

When transmitting `Date`s to clients through the HTTP API, the objects get coerced into a `number` anyway, as JSON does not have a representation for `Date`s. But if we need to do operations on these dates on the server before doing that transmission, or if we want to do things unrelated to transmitting them (e.g. set a timeout for the time between the current date and the one stored in the database, taking time zones into account), having the actual `Date` object is important.

So in short, I definitely think the thing in `MatchmakingTime` should be a `Date`, and you should not be doing this type coercion here.

(and as a side note, since some of your comment here leads me to believe you might be confused on this: `Date` is not a formatted date string, it's an object based around a timestamp + time zone. It can be compared to other dates properly (a later date is `>` an earlier date, for instance), you can subtract 2 dates and get back a delta in milliseconds, etc.)

Reply:
Oh yeah, I misunderstood that. I was thinking of what we send over HTTP back to the clients. But even in that case, I wasn't aware of `Date`'s incompatibility with JSON ^^.

But yeah, all of that makes sense so I'll do it like that. Btw, the only reason why I even did this type coercion here is because we did it before on all of the things that are saved as `Date` in the database, afaik. I never really thought about it much, but might be useful to rethink those use cases as well.

Reply:
I'm not sure what exactly you're looking at, because on a quick look I can only find things that are using Dates as Dates. (`users`, `games`, `chat-channels`, for example, all use Dates the way I described). Perhaps you wrote some code like this a while back and we didn't catch it then?

Reply:
Nvm, you're right. I was thinking of stuff we do in HTTP/WS APIs, like here:
https://github.com/ShieldBattery/ShieldBattery/blob/master/server/lib/wsapi/chat.js#L149
https://github.com/ShieldBattery/ShieldBattery/blob/master/server/lib/api/bans.js#L49

According to the above, this is not really necessary either since those coercions will be done automatically because of JSON's incompatibility with the `Date` object?

Reply:
I checked just to be sure and I was actually slightly wrong on the conversion. They *do* get converted when JSON.stringify'd, but to a standardized string in UTC (e.g. `'2020-12-17T01:05:53.951Z'`). This is probably slightly better than the number/timestamp form, since you can still use this to initialize a `Date` object and the format is self-documenting about the time zone, but probably isn't safe to change code that is currently doing the number conversion directly without looking at what's using it.

Reply:
After actually using it now, it's pretty annoying that you can't use this string directly in the time-related APIs after receiving it on the client, and you pretty much _have_ to convert it to a `Date` object first (preferably immediately in the reducer). That was a nice benefit of using the numbers format imo, since they can be transferred through JSON without any type conversion.